src/hotspot/share/opto/parse.hpp

```diff
@@ -603,4 +603,55 @@ class Parse : public GraphKit {
 #endif
 };
 
+// Specialized uncommon_trap of unstable_if, we have 2 optimizations for them:
+//   1. suppress trivial Unstable_If traps
+//   2. use next_bci of _path to update live locals.
+class UnstableIfTrap {
+  CallStaticJavaNode* const _unc;
+  // Parse::_blocks outlive Parse object itself.
+  // They are reclaimed by ResourceMark in CompileBroker::invoke_compiler_on_method().
+  Parse::Block* const _path; // the pruned path
+  bool _modified;            // modified locals based on next_bci()
+
+public:
+  UnstableIfTrap(CallStaticJavaNode* call, Parse::Block* path): _unc(call), _path(path), _modified(false) {
+    assert(_unc != NULL && Deoptimization::trap_request_reason(_unc->uncommon_trap_request()) == Deoptimization::Reason_unstable_if,
+          "invalid uncommon_trap call!");
+  }
+
+  // The starting point of the pruned block, where control should go
+  // when deoptimization does happen.
+  int next_bci() const {
+    return _path == nullptr ? -1 : _path->start();
+  }
+
+  bool modified() const {
+    return _modified;
+  }
+
+  void set_modified() {
+    _modified = true;
+  }
+
+  Parse::Block* path() const {
+    return _path;
+  }
+
+  // This can only be determined in parse-time.
+  // if _path has only one predecessor, it is trivial if this block is small(1~2 bytecodes)
+  // or if _path has more than one predecessor and has been parsed, _unc does not mask out any real code.
+  bool is_trivial() const {
```

Comment:
But these properties are not checked by the method, right?

Also, the code is only used in debug, should it be guarded?

Reply:
> Looks good overall. Some comments/questions:
> 
>     * Why can't we remove traps that have been modified?
>
In previous revision, I did remove them. Vladimir discovered a special case in tier2.  C2 postponed to do fold-compares until IGVN2. Even though this corner case has been solved in JDK-8287840, I think it's good idea to leave it as a fallback.

 
>     * I'm wondering how useful `Compile::print_statistics()` really is. Is it worth extending it? Is anyone using it?
> 
fair enough. I used them to see how many unstable_if traps are 'trivial'. I think I can remove them now.

>     * Do you need to check for unstable if traps in `Node::destruct`?

thanks for the head-up. technically speaking, yes. in reality, I don't think we call "Node::destruct" for a uncommon_trap.  I will patch it up.